kotlinx-coroutines-core/native/src/MultithreadedDispatchers.kt

```diff
@@ -63,28 +64,54 @@ internal class WorkerDispatcher(name: String) : CloseableCoroutineDispatcher(),
     }
 }
 
-private class MultiWorkerDispatcher(name: String, workersCount: Int) : CloseableCoroutineDispatcher() {
+private class MultiWorkerDispatcher(
+    private val name: String,
+    private val workersCount: Int
+) : CloseableCoroutineDispatcher() {
     private val tasksQueue = Channel<Runnable>(Channel.UNLIMITED)
-    private val workers = Array(workersCount) { Worker.start(name = "$name-$it") }
-
-    init {
-        workers.forEach { w -> w.executeAfter(0L) { workerRunLoop() } }
-    }
+    private val workers = atomicArrayOfNulls<Worker>(workersCount)
+    private val activeWorkers = atomic(0)
 
     private fun workerRunLoop() = runBlocking {
         for (task in tasksQueue) {
-            // TODO error handling
+            /**
+             * Any unhandled exception here will pass through worker's boundary and will be properly reported.
+             */
             task.run()
         }
     }
 
     override fun dispatch(context: CoroutineContext, block: Runnable) {
-        // TODO handle rejections
-        tasksQueue.trySend(block)
+        if (activeWorkers.value != workersCount) {
+            tryAddWorker()
```

Comment:
I don't understand this whole approach. So, each `MultiWorkerDispatcher` has the target number of workers. On each dispatch, we check if this number was reached, and if not, we create a new worker—even if all the existing ones are idle. So, in effect, several dispatches later, the `workersCount` number of workers will be reached. So, why bother at all and not just spawn all the workers immediately, as was done before? I guess the initialization phase will be faster, but is it by much?

Reply:
>I guess the initialization phase will be faster, but is it by much?

To avoid paying the cost of the initialization by the code that does not use it.
E.g. I'm going to introduce `Dispatchers.IO` with 64 workers capacity. We definitely do not want to start 64 workers for the very first use of `Dispatchers.Default` or `Dispatchers.Unconfined`.

If I find a spare time, I'll try to introduce both idle-check and withTimeout-based idle worker termination prior to 1.8.0. 
The former is likely to happen, the latter is unlikely to happen tho.